### DISCUSSION TITLE
Support for STM32F765VGT6

### DISCUSSION THREAD

**Discussion by ghost** [grblHAL/STM32F7xx#16] [Ideas]
I am also looking at the STM32F765VGT6 (100 pin, 1M Flash, 512K RAM, 216MHz). I assume that it will be fairly easy to add support for this specific MCU with Ethernet, WiFi compatability, allow WebUI and to use HSI / LSI clocks?

Cheers
George

**Comment by terjeio** [comment]
Again, basic support will be to add code to set up the [clock tree](https://github.com/grblHAL/STM32F7xx/blob/260af5fecddf7c49c01a584b3186258e2db80f4f/Src/main.c#L46-L112) and perhaps add a few other things like linker script and startup code. I usually do this by creating a new project for the processor in the STM32CubeIDE and copy the needed bits over. So fairly easy.
The on-chip ethernet peripheral is already supported, Wiznet is not, again code can likely be copied - this time from the F4xx driver.
WiFi compatibility is an unknown - I have no idea on how much work that would be. Worst case the whole networking plugin has to be replaced and I am not prepared to do that. WiFi for telnet only (via UART) should be fairly easy - the needed code is likely to be similar to the HC-05 Bluetooth plugin since it also uses AT commands.

**Comment by ghost** [comment]
Thanks for the update.
I will just need to work out the hardware for Ethernet but if you could add the software support that would be great.

I was thinking for WiFi that it could be via SPI but maybe via UART would be simpler to setup. It will be easy to breakout one of the UART ports for WiFi as I believe this MCU also supports USB. Is that something you can add as well or should I use my own USB - UART interface? I read your comments for the STM32F446 discussion and need to take into consideration unique GPIO numbers for each input (interrupt). I guess I will be limited to 16 inputs (Limit, Control, peripherals and AUX In)

**Reply by terjeio** [reply]
> I was thinking for WiFi that it could be via SPI but maybe via UART would be simpler to setup

It sure will, with luck the HC-05 driver code could more or less be reused as-is.

> I believe this MCU also supports USB. Is that something you can add as well or should I use my own USB - UART interface?

I supports USB and the driver does as well. So a USB - UART interface is not needed. Anyway, break out as many [UART ports](https://github.com/grblHAL/STM32F7xx/blob/f87dd92742e65c5b93214b6a38c794958c89218e/Inc/serial.h#L26-L34) as you can - at least two. RS485, pendant, bluetooth/wifi...

> I guess I will be limited to 16 inputs

Correct, except aux in pins does not have to be interrupt capable.

An I2C port together with an interrupt capable pin might be nice to have, at least consider adding an EEPROM/FRAM footprint. The flash page sizes are pretty large in this MCU so writing settings etc. to flash will be rather slow.

**Comment by ghost** [comment]
Awesome.

I have been playing around with STM32CubeIDE and noticed that when I enabled ETHERNET and USB_OTG_FS there was a clock conflict that the IDE could not resolve.

I think then that just using an external USB-UART will be the safetest option. I am thinking USART 1 & 6 or UART 7 & 8 as the pins for both as they are right next to each other (makes it easier for running PCB tracks). What would you suggest is the better option?

For WiFi via UART, I assume there will be a requirement for an Interupt?

**Reply by terjeio** [reply]
> I have been playing around with STM32CubeIDE and noticed that when I enabled ETHERNET and USB_OTG_FS there was a clock conflict that the IDE could not resolve.

Then you have to resolve it manually, I do it nearly all the time. Sometimes I have to accept a MCU master clock frequency slightly less than max allowed.

> What would you suggest is the better option?

Use USB, it is by default faster that UART and that may affect laser engraving speed.

> For WiFi via UART, I assume there will be a requirement for an Interupt?

I guess so, at least it is required for the HC-05 module where a connection is signalled via an interrupt capable auxillary input.

**Reply by ghost** [reply]
OK. Thanks for the feedback.

Would you kindly advise what config options you select for USB_OTG_FS so I can then create the PCB design.

I will also need to run an external crystal (HSE) to get the USB_OTG_FS to work.

I will leave a spare GPIO for the UART channel in case I need it for interrupt.

I will keep checking regularly to see when the build files for this MCU can be downloaded.

_Please let me know if you would like me to upload any config files (from CubeIDE) that you think will help with setting default pin mappings (for what I am currently setting up the PCB design)_

Regards and have a great weekend
George

**Comment by wolfgangr** [comment]
I'd support pledge for the option of adding wiznet `w5100` support to  `STM32F765VGT6`. Just tinkering with a cheap chinese breakoutbard board overe there https://github.com/dresco/STM32H7xx/issues/23. For small lot sizes and breadboard testing, I think that is the easier way to ethernet than using the onchip MAC.

The even better way were to design a nearly bare breakout with ethernet, usb, sd, flash on it. Somebody of you in the pipeline of such a thing?

**Reply by terjeio** [reply]
That basically can be done by copying the low level code over from the F4xx driver and (re)test. BTW it is not a good idea to have ethernet and sdcard sharing the same SPI bus - it is workable but file transfer throughput will be severely limited, possibly by a factor of ten or more.

**Comment by ghost** [comment]
Hi Team,

I have a conceptual design based on default pin mappings obtained from
STM32CubeIDE, (Ethernet via LAN8720 interface IC, Onboard USB, USART 6 for
Future WiFi, SPI [4] for SD Card). Is it possible to create the relevant
build / linker files with these as minimum defaults? I can also provide a
full Map file if that also helps.

Cheers and have a great day
George

**Reply by terjeio** [reply]
Zip your main.c and add to a comment and I'll give it a try.
Note that it is the [LAN8742](https://github.com/grblHAL/STM32F7xx/tree/main/Drivers/BSP/Components/lan8742) that is supported for ethernet - I do not know if that is compatible with the 8720.

**Reply by terjeio** [reply]
I have just commited an update with tentative F765 support (for STM32CubeIDE).

**Reply by ghost** [reply]
Thanks for the update and support.

I will download updates and see how it goes with my custom settings.

Cheers and have a great day / weekend
George

**Comment by ghost** [comment]
[main.c.zip](https://github.com/grblHAL/STM32F7xx/files/14515353/main.c.zip)

Here it is.

The LAN8742 IC has now been added to the design.

Cheers and have a great day
George

**Comment by ghost** [comment]
Also, what is the possibility of adding support for STM32H746ZIT6 MCU?

It looks to closely replicate the STM32F7xxxx based MCU's except that it runs up to 480Mhz clock speed.

If you found the F765VGT6 main.c file useful I can zip and send the H746ZIT6 main.c file as well.

Cheers
George

**Reply by dresco** [reply]
There is a H7 driver [here](https://github.com/dresco/STM32H7xx). 

Currently supports H743 and H723 MCU's (the README needs an update to better reflect the supported boards). You can also see what's supported in the [Web Builder](http://svn.io-engineering.com:8080/).

**Reply by ghost** [reply]
Thanks dresco.

I am happy to provide any files that have already been generated within STM32CubeIDE (specific to Ethernet, USB-OTG-HS, UART and SPI ports) if there is a chance you could create the relvant build / linker files. Happy to maintain this conversation into your repo as well if you wish.

Also my current design is using the H743ZIT6 MCU (typo above). 

Cheers
George

**Comment by ghost** [comment]
Hi terje io,

Apologies for the late reply, I have tried to build and it has thrown 3
errors and 2 warnings:

Description Resource Path Location Type
make: *** [webui/subdir.mk:43: webui/commands_v3.o] Error 1 grblHAL Driver
STM32F756 C/C++ Problem
make: *** Waiting for unfinished jobs.... grblHAL Driver STM32F756 C/C++
Problem
fatal error: fatfs/src/ff.h: No such file or directory sdcard.h /grblHAL
Driver STM32F756/sdcard line 49 C/C++ Problem

Invalid project path: Missing project folder or file: /grblHAL Driver
STM32F756/keypad for Source path. grblHAL Driver STM32F756 pathentry Path
Entry Problem
Invalid project path: Missing project folder or file: /grblHAL Driver
STM32F756/oled for Source path. grblHAL Driver STM32F756 pathentry Path
Entry Problem

Cheers and have a great weekend
George

**Reply by terjeio** [reply]
> fatal error: fatfs/src/ff.h: No such file or directory sdcard.h /grblHAL

The sdcard folder is empty/missing?

> Invalid project path: Missing project folder or file: /grblHAL Driver
STM32F756/keypad for Source path. grblHAL Driver STM32F756 pathentry Path
Entry Problem

The keypad folder is empty/missing?

Missing files/empty folders could be due to the [submodules not beeing downloaded correctly](https://github.com/grblHAL/core/wiki/Compiling-grblHAL).

> Invalid project path: Missing project folder or file: /grblHAL Driver
STM32F756/oled for Source path. grblHAL Driver STM32F756 pathentry Path

My bad, I experimented with OLED support a while back and have left a reference to the code in the configuration, delete it here:

![image](https://github.com/grblHAL/STM32F7xx/assets/20260062/3446704a-0076-40c0-bf0f-fb0e3f4482b4)

**Comment by ghost** [comment]
Thanks for your reply.

I have followed your instructions and deleted the OLED reference and have
created a Keypad folder and populated it with plugin code. I have also
re-downloaded the SD Card plugin code but I still get a SD Card error.

I have attached a few screenshots for your reference.

Regards and have a great day
George

**Reply by terjeio** [reply]
There are no screenshots visible for me.

**Reply by ghost** [reply]
Apologies, I replied via email and did not atach.

Here they are.

Cheers
George
<img width="232" alt="SD card" src="https://github.com/grblHAL/STM32F7xx/assets/57690555/3869c651-ad97-4633-b81b-b44c9e8c0c54">


<img width="716" alt="SD card error" src="https://github.com/grblHAL/STM32F7xx/assets/57690555/7f1eadaa-9995-4058-b68e-1cb6fd6f7c9f">

**Reply by terjeio** [reply]
Oops `STM32_PLATFORM` should be defined, I forgot about that. Add  `|| defined(STM32F765xx)` after `defined(STM32F756xx`) [here](https://github.com/grblHAL/core/blob/69f91b19528c66ba79896e18ffbbecb6d8a2233c/platform.h#L25).

**Reply by ghost** [reply]
Thanks. Fixed that one but has thrown another error.

Cheers
George

<img width="448" alt="My_Machine_Map" src="https://github.com/grblHAL/STM32F7xx/assets/57690555/b70302af-ba49-498e-93d4-8265357b16fe">



<img width="953" alt="SPIN Error" src="https://github.com/grblHAL/STM32F7xx/assets/57690555/935c3c73-90f8-4db1-a632-86a60aeef41e">



<img width="953" alt="Spin En Error" src="https://github.com/grblHAL/STM32F7xx/assets/57690555/28577a3b-5aeb-4675-9fc0-c7981f24af3b">

**Reply by terjeio** [reply]
Spindle pins are now assigned from the aux outputs pool so you have to update your map file accordingly. [An example](https://github.com/grblHAL/STM32F7xx/blob/a26277c75d48a29f97a2c5b8fd57f2771554dc3e/Inc/reference_map.h#L146-L175).

**Reply by ghost** [reply]
Awesome and thanks.

No errors SPINDLE.

BTW how many AUX ports can be assigned?

Cheers and have a great day
George

**Reply by ghost** [reply]
Is it also possible to add support for SPI Port 4?

I have mapped that port as it makes the most sense from a PCB layout perspective.

Cheers George

**Reply by terjeio** [reply]
> BTW how many AUX ports can be assigned?

Aux [inputs](https://github.com/grblHAL/STM32F7xx/blob/a26277c75d48a29f97a2c5b8fd57f2771554dc3e/Src/driver.c#L149-L179) and [outputs](https://github.com/grblHAL/STM32F7xx/blob/a26277c75d48a29f97a2c5b8fd57f2771554dc3e/Src/driver.c#L329-L362) currently supported by the driver. More can be added and some are claimed by the driver.

> Is it also possible to add support for SPI Port 4?

Yes, will add in the next commit.

**Reply by ghost** [reply]
Perfect, Thank you

**Reply by terjeio** [reply]
New commit with SPI4 support is now up.

**Reply by ghost** [reply]
Awesome.

I just updated the files that were changed and tried to build and these errors were thrown:

ERRORS:
Description	Resource	Path	Location	Type
first defined here	grblHAL Driver STM32F756		line 0	C/C++ Problem

Description	Resource	Path	Location	Type
first defined here	grblHAL Driver STM32F756		line 0, external location: /Applications/STM32CubeIDE.app/Contents/Eclipse/plugins/com.st.stm32cube.ide.mcu.externaltools.gnu-tools-for-stm32.11.3.rel1.macos64_1.1.100.202310310803/tools/arm-none-eabi/bin/ld: ./sdcard/sdcard.o	C/C++ Problem

Description	Resource	Path	Location	Type
make: *** [makefile:107: grblHAL Driver STM32F756.elf] Error 1	grblHAL Driver STM32F756		 	C/C++ Problem


Cheers and have a great day
George

**Reply by terjeio** [reply]
What is "first defined here"? It should be named, e.g. a function. BTW I do not get this error, perhaps you should try to clean the project, or build without SD card support to see if it goes away.

**Reply by ghost** [reply]
Perfect.

I only needed to add the keypad and oled folders and then it was completely error and warning free.

Do I need to add a folder for the Trinamic plugin?

Cheers and have a great day
George

**Reply by terjeio** [reply]
> Do I need to add a folder for the Trinamic plugin?

Not unless you want to use Trinamic drivers.

**Reply by ghost** [reply]
Gold.

Thank you for all your support.

Regards
George

**Comment by ghost** [comment]
Just quickly, do the Probe and Safety Door also need to be assigned to AUX (Input) ports?

Cheers George

**Reply by terjeio** [reply]
Yes, when not needed they are not claimed leaving the pins free for general use. Bind them among the highest numbered aux pins.

**Comment by ghost** [comment]
Would you kindly advise if you can add support for I2C4?

I2C4-SDA - GPIO D13
I2C4-SCL - GPIO D12
I2C STROBE - GPIO D11

I have also been enabling more options (in my_machine.h file) and was curious if you are interested in the warnings that have been thrown.

Regards George

**Reply by terjeio** [reply]
try this, not tested:
[i2c.zip](https://github.com/grblHAL/STM32F7xx/files/15044532/i2c.zip)

> I have also been enabling more options (in my_machine.h file) and was curious if you are interested in the warnings that have been thrown.

Yes and no... Newer versions of the compiler is emitting warnings that are patently wrong. And I have fixed some (unused variables).

**Reply by ghost** [reply]
OK. New file seems to have allowed code to compile (with #define I2C_PORT 4)

I then tried to enable Keypad and then it threw an error:

Description	Resource	Path	Location	Type
#error Keypad plugin not supported!	driver.h	/grblHAL Driver STM32F756/Inc	line 284	C/C++ Problem

I do have Strobe defined but is greyed out

#if KEYPAD_ENABLE
#define KEYPAD_PORT             GPIOD
#define KEYPAD_STROBE_PIN       11
#endif


I then changed to 

#if KEYPAD_ENABLE
#define STROBE_PORT      GPIOD
#define STROBE_PIN       11
#endif

I did notice that the STROBE is via an AUXINPUT so I added it in but still get the same error

I suspect I may not have correctly written this for the STROBE.

Cheers and have a great weekend
George
